sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AdaptingIntegerArray.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * An integer array that automatically expands its memory consumption (via copy/allocation) when
+ * reaching limits. This assumes counts remain low, to lower memory overhead.
+ *
+ * <p>This class is NOT thread-safe. It is expected to be behind a synchronized incrementer.
+ *
+ * <p>Instances start by attempting to store one-byte per-cell in the integer array. As values grow,
+ * this will automatically instantiate the next-size integer array (byte => short => int => long)
+ * and copy over values into the larger array. This class expects most usage to remain within the
+ * byte boundary (e.g. cell values < 128).
+ *
+ * <p>This class lives in the (very) hot path of metric recording. As such, we do "fun" things, like
+ * switch on markers and assume non-null based on presence of the markers, as such we suppress
+ * NullAway as it can't understand/express this level of guarantee.
+ *
+ * <p>Implementations MUST preserve the following:
+ *
+ * <ul>
+ *   <li>If cellSize == BYTE then byteBacking is not null
+ *   <li>If cellSize == SHORT then shortBacking is not null
+ *   <li>If cellSize == INT then intBacking is not null
+ *   <li>If cellSize == LONG then longBacking is not null
+ * </ul>
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public class AdaptingIntegerArray {
+  @Nullable private byte[] byteBacking;
+  @Nullable private short[] shortBacking;
+  @Nullable private int[] intBacking;
+  @Nullable private long[] longBacking;
+
+  /** Possible sizes of backing arrays. */
+  private enum ArrayCellSize {
+    BYTE,
+    SHORT,
+    INT,
+    LONG
+  }
+  /** The current byte size of integer cells in this array. */
+  private ArrayCellSize cellSize;
+
+  /** Construct an adapting integer array of a given size. */
+  public AdaptingIntegerArray(int size) {
+    this.cellSize = ArrayCellSize.BYTE;
+    this.byteBacking = new byte[size];
+  }
+
+  private AdaptingIntegerArray(int size, ArrayCellSize cellSize) {
+    this.cellSize = cellSize;
+    switch (cellSize) {
```

Comment:
Can you add tests for the missed lines in this class, or if they don't have a short term need maybe remove the unused methods? There are some methods in other classes too

Reply:
Will do.   I think I added these in preparation of preserving cell-size between collection intervals, I'll either add tests now or move that to a seperate branch/PR fully.